Seperate internal metrics declaration model from the v1 (de)serialization model

adamconnelly opened the pull request:
Work towards #513

## Proposed Changes

This PR contains changes to create a separation between the internal model used by promitor, and the objects used for serialization / deserialization. I've split it into a few commits to make it a bit easier to understand:

- The first commit just duplicates the existing model objects, and puts them in the `Promitor.Core.Scraping.Configuration.Serialization.v1` namespace.
- The second commit converts the v1 models into "Builders" that can be used to create the internal models.
- The last two commits remove an unnecessary method, and update the tests to work with the new model.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tomkerkhove/promitor?pullRequest=652) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tomkerkhove/promitor?pullRequest=652) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tomkerkhove/promitor?pullRequest=652) it.</sub>

adamconnelly commented:
@tomkerkhove I notice the CodeFactor check is failing because of the duplicate ResourceType enum. I've actually removed the duplicate in a separate branch where I started working on changing the internal model to support multiple resources. I can alter this branch to remove it, but if you don't mind the check failing temporarily until the next stage is complete it'll just make my life slightly simpler when I go to rebase stuff together.

Just let me know.

tomkerkhove commented:
Thanks for the PR @adamconnelly!

I've quickly gone through it to get an initial understanding and let me know if this is correct:
- `Promitor.Core.Scraping.Configuration.Model.Metrics.ResourceTypes.{Type}MetricDefinition` is used internally by the runtime, just like before
- `Promitor.Core.Scraping.Configuration.Serialization.v1.Model.Metrics.ResourceTypes.{Type}MetricDefinitionBuilder` is used to deserialize the configuration into our internal model, consumed via our v1 deserializers

Is that correct?

Frankly, the builders feel a bit odd given they are not really following the builder pattern but I'm even wondering if we need them. I'd either use the v1 deserializers to map into a v1 model and do the translation there, or map from v1 deserializer directly to the internal model. Thoughts?

A nice to have would be to remove the default constructor and/or setters on our `Promitor.Core.Scraping.Configuration.Model.Metrics.ResourceTypes.{Type}MetricDefinition` in order to stimulate using the more specific constructors. What do you think? (less urgent)

adamconnelly commented:
@tomkerkhove that's totally correct - I've basically just taken a copy of the model objects, put them under the v1 namespace, and renamed them. So in theory the runtime is unchanged, but the deserialization code gets its own set of classes that exactly match the v1 format.

> Frankly, the builders feel a bit odd given they are not really following the builder pattern but I'm even wondering if we need them.

I agree, it doesn't feel completely right. I wonder if that could be solved by figuring out a more appropriate name for them though? I think it's helpful for them to have different names to the internal model because then you don't end up confused between the two sets of classes.

This is maybe overkill, but I guess we could prefix or suffix them with their format version, so `V1MetricDefinition` or `MetricDefinitionV1` for example?

> I'd either use the v1 deserializers to map into a v1 model and do the translation there, or map from v1 deserializer directly to the internal model. Thoughts?

So true confession - when I originally pushed the branch up I hadn't completely understood the codebase, and I'd forgotten to update the `v1.Core.{Type}Deserializer` objects to return the serialization objects, and they were returning the internal model by accident. This actually worked completely fine, so we could take this approach.

My only slight misgiving about it is that it feels slightly wrong. We then end up in a situation where the serialization objects are actually only used for serialization, and not deserialization.

I feel like we should aim for:

1. Read yaml file.
2. Figure out version and create the correct serializer.
3. Deserialize to the version-specific model.
4. Convert to the internal model.

Maybe if we renamed the `Build()` methods to `CreateModel()` or something like that?

> A nice to have would be to remove the default constructor and/or setters on our Promitor.Core.Scraping.Configuration.Model.Metrics.ResourceTypes.{Type}MetricDefinition in order to stimulate using the more specific constructors. What do you think? (less urgent)

Sure - I didn't do that as part of my changes because I didn't want to make too many changes to the existing code but it makes complete sense to me. Do you mind if I do that in a separate PR after we've got my work on this issue merged? I'm just trying to juggle a few branches at the moment to keep different parts of the work separate, and it's a little painful resolving conflicts with myself 😸

tomkerkhove commented:
> My only slight misgiving about it is that it feels slightly wrong. We then end up in a situation where the serialization objects are actually only used for serialization, and not deserialization.

Isn't it rather the other way around? Only used for deserialization and not for serialization? The reason for that is that it uses the types and we need to write our own deserialization given the inheritance etc and could not get it to work OOTB. 
Serialization, however, works OOTB so I'm using that approach.

Or are you talking about another aspect? Or is it given we've introduced the version? If so, maybe we should track this conversation in a dedicated issue if it's not required for #513.

> I agree, it doesn't feel completely right. I wonder if that could be solved by figuring out a more appropriate name for them though? I think it's helpful for them to have different names to the internal model because then you don't end up confused between the two sets of classes.
> 
> This is maybe overkill, but I guess we could prefix or suffix them with their format version, so V1MetricDefinition or MetricDefinitionV1 for example?

How about the following then?
1. Deserializers (ie v1) maps to v1 model
2. Mappers map from v1 to internal model
3. Runtime uses internal model

We can then rename the builders to `{type}Mapper`, or something similar, which has a `MapToInternalFormat` method?

> Sure - I didn't do that as part of my changes because I didn't want to make too many changes to the existing code but it makes complete sense to me. Do you mind if I do that in a separate PR after we've got my work on this issue merged? I'm just trying to juggle a few branches at the moment to keep different parts of the work separate, and it's a little painful resolving conflicts with myself 😸

That's certainly ok for me, thanks!

adamconnelly commented:
> Isn't it rather the other way around? Only used for deserialization and not for serialization? The reason for that is that it uses the types and we need to write our own deserialization given the inheritance etc and could not get it to work OOTB.

So what I was meaning here, is that if the deserializers return the runtime model rather than the v1 model, we end up with a situation where the v1 model only gets used for serialization, but we don't populate it during deserialization.

I think ideally the V1Deserializer (I've just made this name up for the sake of this discussion) should return the V1 model, and the V2Deserializer should return the V2 model. We then map from the version specific model to the runtime model like you suggest.

That's actually what the code in this PR does, but I've maybe blurred things a bit by naming the V1 model objects "builders", and having them handle mapping to the runtime model.

So to get to where I think you want it, what I'd need to do is:

1. Rename `{Type}Builder` to `{Type}V1`.
2. Remove the Build() methods from them, and create a separate Mapper object.

I know I was a bit wary about using AutoMapper, but after thinking about it more I think it probably makes sense.

Before I make any changes, does that make sense to you, and are you ok with me adding AutoMapper as a dependency?

tomkerkhove commented:
> So what I was meaning here, is that if the deserializers return the runtime model rather than the v1 model, we end up with a situation where the v1 model only gets used for serialization, but we don't populate it during deserialization.
> 
> I think ideally the V1Deserializer (I've just made this name up for the sake of this discussion) should return the V1 model, and the V2Deserializer should return the V2 model. We then map from the version specific model to the runtime model like you suggest.

That makes sense indeed, but think it would make sense to have one (de)serilizer that orchestrates this based on the version field and obfuscates this from the consumer, I don't know if that makes sense to you?

> That's actually what the code in this PR does, but I've maybe blurred things a bit by naming the V1 model objects "builders", and having them handle mapping to the runtime model.
> 
> So to get to where I think you want it, what I'd need to do is:
> 
> 1. Rename `{Type}Builder` to `{Type}V1`.
> 2. Remove the Build() methods from them, and create a separate Mapper object.

In that case, can't we just move this to our v1 configuration model that we already have or am I missing something?

The (de)serializers can use that to create the v1 model, return it and the more "general" (de)serializer can do the mapping to the internal format.

However, I think we're on the same page here but you're just suggesting to suffix the model classes with V1. Given it's already in the namespace it feels a bit like overkill but can't hurt - Go ahead!

> I know I was a bit wary about using AutoMapper, but after thinking about it more I think it probably makes sense.
> 
> Before I make any changes, does that make sense to you, and are you ok with me adding AutoMapper as a dependency?

I'm certainly ok with adding a dependency with AutoMapper, it will make it a lot easier for us! The only requirement I would have is to have a unit test per mapping scenario to avoid breaking maps later on. Sounds good to you?

Thanks again mate!

adamconnelly commented:
@tomkerkhove I've added some more commits to this that hopefully get it to where you're wanting it. We still have a single point that handles the deserialization based on the `version` field. The only thing I've changed around that is that the v1 `ConfigurationSerializer` class returns a `MetricsDeclarationV1` instead of a `MetricsDeclaration`. The other (non-version specific?) `ConfigurationSerializer` object handles the mapping and returns a `MetricsDeclaration`.

Here's a rough summary of the commits:

1. Renames the v1 configuration objects from `{Type}Builder` to `{Type}V1`. I thought about not including the V1 suffix, but it actually makes things a lot more sane when you go to write the AutoMapper profiles.
2. Add AutoMapper, configure the mapper profiles, and add a test to verify the mappings are valid. I haven't added tests for each individual object being mapped because at the moment there's no custom mappings. If we rename a property in future (for example), the general AutoMapper test I've added will fail.
3. Fixed the tests. I kept this separate because the volume of changes was huge.

Let me know what you think of that.

Also, on a side-note - I signed the CLA a few days ago but that check is still failing. Any idea what the problem is?

adamconnelly commented:
Just noticed I forgot to actually remove the `Build()` methods - I've pushed another commit with that change.

tomkerkhove commented:
Thanks @adamconnelly, I'll give it another review!

The CLA Assistant is failing because you've signed it with @adamconnelly 	while the commits you did are done via @adam-resdiary 	. Once you sign it with @adam-resdiary 	 instead you should be good to go!

adamconnelly commented:
🤦‍♂ I should have spotted that. I'll try to sort that. I've got a separate work account to try to keep things separate, and I wanted to make the changes with my personal account. I must not have set it up properly.

tomkerkhove commented:
That's ok, no worries! Feel free to squash commits with your personal account if you want, PRs get squashed anyway so it's fine for me.

adamconnelly commented:
Ok - that's it sorted now.

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs:
```diff
@@ -4,6 +4,26 @@ namespace Promitor.Core.Scraping.Configuration.Model.Metrics
+            Dictionary<string, string> labels,
+            Scraping scraping)
+        {
+            AzureMetricConfiguration = azureMetricConfiguration;
```
I don't think we want to create invalid metric definitions so suggest to guard for invalid values. Thoughts?

adamconnelly replied:
You just thinking of adding `Guard.NotNull()` calls here? If so makes perfect sense to me.

tomkerkhove replied:
I think that should be good enough for now, for strings you should be able to use NotNullOrEmpty or something similar

tomkerkhove replied:
Resolving this conversation as it's no longer applicable.

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs:
```diff
@@ -4,6 +4,26 @@ namespace Promitor.Core.Scraping.Configuration.Model.Metrics
+        {
+        }
+
+        protected MetricDefinition(
```
Nitpicking - Think name, description, resourceGroupName, labels, scraping, azureMetricConfiguration would make a more logical ordering

adamconnelly replied:
Sure - easy to sort.

tomkerkhove replied:
Thanks!

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ResourceTypes/ContainerInstanceMetricDefinition.cs:
```diff
@@ -1,7 +1,19 @@
+
+        public ContainerInstanceMetricDefinition(AzureMetricConfiguration azureMetricConfiguration, string description, string name, string resourceGroupName, string containerGroup, Dictionary<string, string> labels, Scraping scraping)
+            : base(azureMetricConfiguration, description, name, resourceGroupName, labels, scraping)
+        {
```
Same question for all properties - Should we enforce valid configurations via guarding or not?

adamconnelly replied:
I think it makes sense so I'll update that. The slight catch here is that I think AutoMapper will be using the zero argument constructors until we remove them (assuming we can). But I can add the guards in here, and when we get rid of the zero argument constructors it'll be enforced correctly.

tomkerkhove replied:
Perfect, thanks!

adamconnelly replied:
@tomkerkhove slight hiccup with adding the Guards in. The problem is that the validation steps rely on being able to create an invalid object, and there are then validation rules to ensure all the required properties have been set.

For example, `AzureStorageQueueMetricsDeclarationValidationStepTests.AzureStorageQueuesMetricsDeclaration_DeclarationWithoutAccountName_Fails` will fail if I guard against the account name being null or empty because it tries to create a `StorageQueueMetricDefinition` with an empty account name.

So I think if we wanted to go down that route we'd need to rethink how the validation works.

What's your thoughts on that?

tomkerkhove replied:
I think it's fine to leave out the guards for now then, it's clear that it's covered for now.

Once we've releases 1.0 we can revise a few things like this

tomkerkhove replied:
Thanks btw!

Resolving this conversation as it's no longer applicable.

tomkerkhove commented on src/Promitor.Scraper.Tests.Unit/Serialization/AutoMapperTests.cs:
```diff
@@ -0,0 +1,19 @@
+            var mapperConfiguration = new MapperConfiguration(c => c.AddProfile<V1MappingProfile>());
+
+            // Act / Assert
+            mapperConfiguration.AssertConfigurationIsValid();
```
Will this ensure that everything maps correctly?

adamconnelly replied:
As things stand I think it will because the model is a one-to-one with the config. So if we renamed the `ResourceGroupName` property on the runtime model to `ResourceGroup` for example, this tests would fail because the property names no-longer match.

In addition, all of the current deserialization tests rely on serializing and then deserializing to the runtime model. As a result of that, they actually test the AutoMapper mappings as well.

Once we've got this PR merged, I'll push up the next stage (altering the runtime model). I've added additional tests for the AutoMapper config to that because there's some parts that need custom mappings to work.

Let me know if that's good enough for you, or whether you want anything else added.

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Serialization/v1/Mapping/V1MappingProfile.cs:
```diff
@@ -0,0 +1,48 @@
+            CreateMap<MetricAggregationV1, MetricAggregation>();
+            CreateMap<SecretV1, Secret>();
+
+            CreateMap<ContainerInstanceMetricDefinitionV1, ContainerInstanceMetricDefinition>();
```
We might want to add this to https://github.com/tomkerkhove/promitor/blob/master/adding-a-new-scraper.md

adamconnelly replied:
I was going to redo this after adding the v2 implementation. Are you ok with it being temporarily out of date, or do you want me to update it as part of this PR?

I don't mind either way, but obviously if I can avoid changing it and then changing it again shortly I'll do that 😸

tomkerkhove replied:
Sure, I'm ok with holding off for v2

adamconnelly commented:
Ok - that should be it sorted now.

Once this is merged, I've got the next stage prepped (changing the runtime model to support multiple resources). It might not be until later in the week before I get a chance to push up a PR for it though because it's going to involve a bit of conflict merging based on some of the changes we've made here.

I've also got the v2 config working with multiple resources, and I quickly did an end to end test to make sure it works. It still needs a bit of work before it's ready for a PR, but shouldn't take too long (I mainly needed to sort some tests that became broken as a result of adding AutoMapper to this PR, and I just had avoided fixing them properly until we got this PR sorted).

tomkerkhove commented:
That's awesome @adamconnelly, thank you for all the contributions!

I'm merging this and feel free to take all the time you need, no worries!

tomkerkhove commented:
@adamconnelly 	For the next part, feel free to just replace our current v1 models as we'll ship v1.0 soon and would like to use the new model as v1.

Is that ok for you?

adamconnelly commented:
That's fine, it might actually make my life a bit easier. It won't be in the next PR though - it'll be one after that. I could try to combine the two sets of changes together, but I think at this stage it'll be safer to keep them separate.

tomkerkhove commented:
Sounds good!
